src/performRequest.js

```diff
@@ -0,0 +1,161 @@
+}
+
+
+/**
+ * Returns the encoding as either 'utf-8' or 'base64'. Throws
+ * an error in case any other encoding is provided.
+ *
+ * @param {string} encoding
+ */
+function normalizeBodyEncoding(encoding) {
+  if (!encoding) {
+    return 'utf-8';
+  }
+  const lcEncoding = encoding.toLowerCase();
+  if (lcEncoding === 'utf-8' || lcEncoding === 'utf8') {
+    return 'utf-8';
+  }
+  if (lcEncoding === 'base64') {
+    return 'base64';
+  }
+  throw new Error(`Unsupported encoding: '${encoding}' (only UTF-8 and `
+    + 'Base64 are supported)');
+}
+
+
+/**
+ * Detects an existing Content-Length header and overrides the user-provided
+ * header value in case it's out of sync with the real length of the body.
+ *
+ * @param {Object} headers HTTP request headers
+ * @param {Buffer} body HTTP request body
+ * @param {Object} [options]
+ * @param {Object} [options.logger] Custom logger
+ */
+function setContentLength(headers, body, options = {}) {
+  const logger = options.logger || defaultLogger;
+
+  const modifiedHeaders = Object.assign({}, headers);
+  const calculatedValue = Buffer.byteLength(body);
+  const name = caseless(modifiedHeaders).has('Content-Length');
+  if (name) {
+    const value = parseInt(modifiedHeaders[name], 10);
+    if (value !== calculatedValue) {
+      modifiedHeaders[name] = `${calculatedValue}`;
+      logger.warn(`Specified Content-Length header is ${value}, but the real `
+        + `body length is ${calculatedValue}. Using ${calculatedValue} instead.`);
+    }
+  } else {
+    modifiedHeaders['Content-Length'] = `${calculatedValue}`;
+  }
+  return modifiedHeaders;
+}
+
+
+/**
+ * Real transaction response object factory. Serializes binary responses
+ * to string using Base64 encoding.
+ *
+ * @param {Object} res Node.js HTTP response
+ * @param {Buffer} body HTTP response body as Buffer
+ */
+function createTransactionRes(res, body) {
+  const transactionRes = {
+    statusCode: res.statusCode,
+    headers: Object.assign({}, res.headers)
+  };
+  if (Buffer.byteLength(body || '')) {
+    transactionRes.bodyEncoding = detectBodyEncoding(body);
+    transactionRes.body = body.toString(transactionRes.bodyEncoding);
+  }
+  return transactionRes;
+}
+
+
+/**
+ * @param {Buffer} body
+ */
+function detectBodyEncoding(body) {
+  // U+FFFD is a replacement character in UTF-8 and indicates there
+  // are some bytes which could not been translated as UTF-8. Therefore
+  // let's assume the body is in binary format. Dredd encodes binary as
+  // Base64 to be able to transfer it wrapped in JSON over the TCP to non-JS
+  // hooks implementations.
+  return body.toString().includes('\ufffd') ? 'base64' : 'utf-8';
+}
+
+
+performRequest.normalizeBodyEncoding = normalizeBodyEncoding;
+performRequest.getBodyAsBuffer = getBodyAsBuffer;
+performRequest.setContentLength = setContentLength;
+performRequest.createTransactionRes = createTransactionRes;
+performRequest.detectBodyEncoding = detectBodyEncoding;
```

Comment:
When we are exposing private/internal APIs for testing purposes, I think it would be good idea to prefix the names with leading underscores `__` - see [this](https://github.com/facebook/react/blob/d17091ddebf317173eb30e3ab4a8e28e124c8f50/src/umd/ReactWithAddonsUMDEntry.js#L22) for inspiration 🙂

Reply:
lol (inspiration). I did some research and dropped some thoughts on this to https://github.com/apiaryio/dredd-transactions/pull/179#discussion_r206897044. Let's continue there.